### PR TITLE
feat(fgw): replacing the reqwest feeder client with hyper

### LINKED ISSUE #192: feat(logs): Improve rate limit logs displaying
### Is there an existing issue?

- [X] I have searched the existing issues

### Motivation

Currently, when a rate limit occurs during the request set to the FGW, we get about ten error logs for each request, cluttering the logs. Due to our highly optimized fetching system, we reach the limit of 200 requests per minute per IP address.

Capture d’écran 2024-07-01 à 20 47 03

### Request

Therefore, we need to find a solution to reduce these logs to a single log with, if possible, more detailed information in the error log.

### Solution

The solution would be to reduce these logs to a single line with a counter for the ongoing requests and a counter for the seconds remaining before a retry. In the following format:

```bash
[2024-07-01 12:25:41 INFO] The fetching process has been rate limited with {retry_number} request. Retrying in {retry_time_seconds}s
```

### Are you willing to help with this request?

Yes!

### PR DESCRIPTION
**Pull request by jbcaron**
# Pull Request type

- Refactoring (no functional changes, no API changes)

## What is the current behavior?

Resolves: #192 

## What is the new behavior?

- 

## Does this introduce a breaking change?

No

## Other information

same as #357

### CODE REVIEW

File: crates/client/gateway/src/client/builder.rs [hunk 1]
```diff
@@ -48,4 +76,170 @@ impl FeederClient {
...
+    pause_until: Arc<RwLock<Option<Instant>>>,
+}
+
+impl RetryPolicy {
+    pub fn new(max_retries: usize, backoff: Duration, pause_until: Arc<RwLock<Option<Instant>>>) -> Self {
+        RetryPolicy { max_retries, backoff, pause_until }
+    }
+}
+
+impl Policy<Request<String>, Response<Incoming>, Box<dyn Error + Send + Sync>> for RetryPolicy {
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/builder.rs]: Consider updating the code to make it more evident this is a retry policy (there also exists `tower::hedge::Policy`)

```suggestion
impl retry::Policy<Request<String>, Response<Incoming>, Box<dyn Error + Send + Sync>> for RetryPolicy {
```

File: crates/client/gateway/src/client/builder.rs [hunk 2]
```diff
@@ -48,4 +76,170 @@ impl FeederClient {
...
+        &self,
+        _: &Request<String>,
+        result: Result<&Response<Incoming>, &Box<dyn Error + Send + Sync>>,
+    ) -> Option<Self::Future> {
+        let pause_until = self.pause_until.clone();
+
+        match result {
+            Ok(response) => {
+                // TOO_MANY_REQUESTS
+                if response.status().as_u16() == 429 {
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/builder.rs]: Consider using `http::StatusCode::TOO_MANY_REQUESTS` instead.

```suggestion
                if response.status() == http::StatusCode::TOO_MANY_REQUESTS {
```

File: crates/client/gateway/src/client/builder.rs [hunk 3]
```diff
@@ -48,4 +76,170 @@ impl FeederClient {
...
+fn get_retry_after(response: &Response<Incoming>) -> Option<Duration> {
+    if let Some(retry_after_header) = response.headers().get("Retry-After") {
+        if let Ok(retry_after_str) = retry_after_header.to_str() {
+            if let Ok(retry_seconds) = retry_after_str.parse::<u64>() {
+                return Some(Duration::from_secs(retry_seconds));
+            }
+        }
+    }
+    None
+}
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/builder.rs:194]: This seems counter-productive: what is the point of rate limiting if the request can pass a `Rety-After` header with a duration of 0s? Imo we should default all requests to a retry duration of 10s.

File: crates/client/gateway/src/client/builder.rs [hunk 4]
```diff
@@ -48,4 +76,170 @@ impl FeederClient {
...
+                    let fut = async move {
+                        let mut pause_lock = pause_until.write().await;
+                        *pause_lock = Some(Instant::now() + retry_after);
+
+                        // wait for the retry_after duration
+                        tokio::time::sleep(retry_after).await;
+
+                        next_policy
+                    };
+                    Some(Box::pin(fut))
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/builder.rs]: Consider using `futures::future::FutureExt` instead. This is more of a nitpick and requires adding
```toml
futures = { workspace = true }
```
to this crates `cargo.toml`.

```suggestion
                    let fut = async move {
                        let mut pause_lock = pause_until.write().await;
                        *pause_lock = Some(Instant::now() + retry_after);

                        // wait for the retry_after duration
                        tokio::time::sleep(retry_after).await;

                        next_policy
                    }.boxed();
                    Some(fut)
```

File: crates/client/gateway/src/client/builder.rs [hunk 5]
```diff
@@ -48,4 +76,170 @@ impl FeederClient {
...
+                let next_policy = RetryPolicy {
+                    max_retries: self.max_retries - 1,
+                    backoff: self.backoff,
+                    pause_until: self.pause_until.clone(),
+                };
+                let fut = tokio::time::sleep(self.backoff);
+                Some(Box::pin(async move {
+                    fut.await;
+                    next_policy
+                }))
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/builder.rs]: Same here.

File: crates/client/gateway/src/client/builder.rs [hunk 6]
```diff
@@ -48,4 +76,170 @@ impl FeederClient {
...
+            };
+
+            if let Some(duration) = pause_duration {
+                tokio::time::sleep(duration).await;
+            }
+
+            inner.call(req).await
+        };
+
+        Box::pin(fut)
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/builder.rs]: Same

File: crates/client/gateway/src/client/request_builder.rs [hunk 7]
```diff
@@ -71,44 +75,74 @@ impl<'a> RequestBuilder<'a> {
         unpack(self.send_get_raw().await?).await
     }
 
-    pub async fn send_get_raw(self) -> Result<Response, SequencerError> {
-        self.client.get(self.url).headers(self.headers).query(&self.params).send().await.map_err(Into::into)
+    pub async fn send_get_raw(self) -> Result<Response<Incoming>, SequencerError> {
+        let uri = self.build_uri()?;
+
+        let mut req_builder = Request::builder().method("GET").uri(uri);
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/request_builder.rs]: Consider using `http::Method::GET` instead.
```suggestion
        let mut req_builder = Request::builder().method(http::Method::GET.as_str()).uri(uri);
```

File: crates/client/gateway/src/client/request_builder.rs [hunk 8]
```diff
@@ -71,44 +75,74 @@ impl<'a> RequestBuilder<'a> {
...
+
+        for (key, value) in self.headers.iter() {
+            req_builder = req_builder.header(key, value);
+        }
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/request_builder.rs]: Something like this should be more performant:
```rust
req_builder.headers_mut().unwrap().extend(self.headers);
```

File: crates/client/gateway/src/client/request_builder.rs [hunk 9]
```diff
@@ -71,44 +75,74 @@ impl<'a> RequestBuilder<'a> {
...
 
     #[allow(dead_code)]
     pub async fn send_post<T>(self) -> Result<T, SequencerError>
     where
         T: DeserializeOwned,
     {
-        let mut request = self.client.post(self.url);
+        let uri = self.build_uri()?;
+
+        let mut req_builder = Request::builder().method("POST").uri(uri);
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/request_builder.rs]: Consider using `http::Method::POST` instead.
```suggestion
        let mut req_builder = Request::builder().method(http::Method::POST.as_str()).uri(uri);
```

File: crates/client/gateway/src/client/request_builder.rs [hunk 10]
```diff
@@ -71,44 +75,74 @@ impl<'a> RequestBuilder<'a> {
...
 
         for (key, value) in self.headers.iter() {
-            request = request.header(key, value);
+            req_builder = req_builder.header(key, value);
         }
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/request_builder.rs]: Same here

File: crates/client/gateway/src/client/request_builder.rs [hunk 11]
```diff
@@ -71,44 +75,74 @@ impl<'a> RequestBuilder<'a> {
...
 }
 
-async fn unpack<T>(response: reqwest::Response) -> Result<T, SequencerError>
+async fn unpack<T>(response: Response<Incoming>) -> Result<T, SequencerError>
 where
     T: ::serde::de::DeserializeOwned,
 {
     let http_status = response.status();
-    if http_status == reqwest::StatusCode::TOO_MANY_REQUESTS {
+    let whole_body = response.collect().await.unwrap().aggregate();
```
Comment (Trantorian1) [inline crates/client/gateway/src/client/request_builder.rs]: Replace this `unwrap` by an `expect`

**Review by Trantorian1 — CHANGES_REQUESTED** [review]

**Comment by Trantorian1** [comment]
Please also consider adding back a message signifying that rate limiting has occurred as otherwise this give the impression Madara is being unresponsive.

**Comment by Trantorian1** [comment]
Also there seems to be some difficulties with `Ctrl-C`ing out of Madara when the node sleeps due to rate limiting.

File: crates/client/gateway/src/client/builder.rs (hunk 3 above)
Comment (jbcaron) [reply crates/client/gateway/src/client/builder.rs:194]: yes we can define a minimum, after verification the sequencer does not send this header

**Comment by jbcaron** [comment]
> Also there seems to be some difficulties with `Ctrl-C`ing out of Madara when the node sleeps due to rate limiting.

fix

**Review by Trantorian1 — APPROVED** [review]
All good!

**Pull request closed without merging**